### PR TITLE
Improve sparse adagrad update

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
- The weight decay term in Adagrad is incorrect. See https://github.com/apache/incubator-mxnet/issues/9363 for more details. 
- The sparse update is not fast enough for the language model I'm training. Moving the logic to cpp backend so that we don't suffer from blocking calls such as `RowSparseNDArray.indices` and `RowSparseNDArray.data`. 

@ZiyueHuang @piiswrong @sxjscience @cjolivier01 @szha 
Commit author is mistaken. Nvm. 

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Feature1, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: python/mxnet/optimizer.py [hunk 1]
```diff
@@ -1075,42 +1083,24 @@ def update(self, index, weight, grad, state):
...
-            state[:] = history
-            assert state.stype == save_history_stype
-            assert len(history_indices) == grad_indices_count
+            if wd != 0.0:
+                raise NotImplementedError("sparse adagrad with weight decay is not implemented")
+            kwargs = {'epsilon': self.float_stable_eps,
+                      'rescale_grad': self.rescale_grad}
+            if self.clip_gradient:
+                kwargs['clip_gradient'] = self.clip_gradient
+            sparse.adagrad_update(weight, grad, history, out=weight, lr=lr, wd=wd, **kwargs)
```
Comment (szha) [inline python/mxnet/optimizer.py]: since wd is still passed to the operator, we can just check whether wd is 0 or not in the backend.

File: src/operator/optimizer_op-inl.h [hunk 2]
```diff
@@ -1484,7 +1484,153 @@ inline void SignumUpdate(const nnvm::NodeAttrs& attrs,
...
 
+struct AdagradParam : public dmlc::Parameter<AdagradParam> {
+  float lr;
+  float epsilon;
+  float rescale_grad;
+  float clip_gradient;
+  float wd;
+  DMLC_DECLARE_PARAMETER(AdagradParam) {
+    DMLC_DECLARE_FIELD(lr)
+    .describe("Learning rate");
```
Comment (szha) [inline src/operator/optimizer_op-inl.h]: indentation?

Comment (eric-haibin-lin) [reply src/operator/optimizer_op-inl.h]: It's indented. What's the concern?

File: python/mxnet/optimizer.py (hunk 1 above)
Comment (eric-haibin-lin) [reply python/mxnet/optimizer.py]: Done

File: src/operator/optimizer_op-inl.h (hunk 2 above)
Comment (szha) [reply src/operator/optimizer_op-inl.h]: nevermind. it's correct

**Comment by cjolivier01** [comment]
Is this ready to go in?

**Comment by eric-haibin-lin** [comment]
Hi everyone,

Considering the `wd` term is implemented inconsistently in many other optimizers (see https://github.com/apache/incubator-mxnet/issues/9881) , I removed the changes for `wd` in this PR so that the scope of changes is limited. Let's discuss how to fix the behavior of `wd` systematically in a separate issue/PR.
